packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-undo-actions-modal.component.tsx

```diff
@@ -0,0 +1,96 @@
+import React, { type ReactNode, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { type QueueEntry } from '../../types';
+import { Button, ModalHeader, ModalBody, ModalFooter, Stack } from '@carbon/react';
+import { type FetchResponse, showSnackbar } from '@openmrs/esm-framework';
+import { useMutateQueueEntries } from '../../hooks/useMutateQueueEntries';
+
+interface QueueEntryUndoActionsModalProps {
+  queueEntry: QueueEntry;
+  closeModal: () => void;
+  modalParams: ModalParams;
+}
+
+interface ModalParams {
+  modalTitle: string;
+  modalInstruction: ReactNode;
+  submitButtonText: string;
+  submitSuccessTitle: string;
+  submitSuccessText: string;
+  submitFailureTitle: string;
+  submitAction: (queueEntry: QueueEntry) => Promise<FetchResponse<any>>;
+}
+// Modal with the same UI for undoing a queue entry transition and voiding a queue entry
+export const QueueEntryUndoActionsModal: React.FC<QueueEntryUndoActionsModalProps> = ({
+  queueEntry,
+  closeModal,
+  modalParams,
+}) => {
+  const { t } = useTranslation();
+  const { mutateQueueEntries } = useMutateQueueEntries();
+  const {
+    modalTitle,
+    modalInstruction,
+    submitButtonText,
+    submitSuccessTitle,
+    submitSuccessText,
+    submitFailureTitle,
+    submitAction,
+  } = modalParams;
+
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const submitForm = (e) => {
+    e.preventDefault();
+    setIsSubmitting(true);
+
+    submitAction(queueEntry)
+      .then(({ status }) => {
+        const success = status >= 200 && status < 300;
+        if (success) {
+          showSnackbar({
+            isLowContrast: true,
+            title: submitSuccessTitle,
+            kind: 'success',
+            subtitle: submitSuccessText,
+          });
+          mutateQueueEntries();
+          closeModal();
+        } else {
+          throw { message: t('unexpectedServerResponse', 'Unexpected Server Response') };
+        }
+      })
+      .catch((error) => {
+        showSnackbar({
+          title: submitFailureTitle,
+          kind: 'error',
+          subtitle: error?.message,
```

Comment:
Ideally you'd work with your back end changes to ensure whatever message is coming from the back end that indicates a failure state (eg. no previous queue entry, multiple matching previous queue entries, etc) could be passed here via a localized message.  i.e. can we create an appropriate Exception class that, when it propogates through to here, we have access to both a message code and a default message that we can use to present the error.